contracts/optimistic-ethereum/OVM/OVM_BondManager.sol

```diff
@@ -0,0 +1,176 @@
+            // cases of fraud for a batch e.g. because of multiple invalid state
+            // transitions? should these be penalized multiple times?
+            bonds[sequencers[batchIndex]] += requiredCollateral;
+        }
+
+        // Reset the storage slot to disallow this from being called multiple times
+        delete sequencers[batchIndex];
+    }
+
+    // Claims the user's proportion of the provided state
+    function payout(bytes32 _preStateRoot) public {
+        Rewards storage rewards = witnessProviders[_preStateRoot];
+
+        // only allow claiming if fraud was proven in `finalize`
+        require(rewards.canClaim, "Cannot claim rewards");
+
+        // proportional allocation
+        uint256 amount = (requiredCollateral * rewards.numClaims[msg.sender]) / rewards.total;
+
+        // reset the user's claims so they cannot double claim
+        rewards.numClaims[msg.sender] = 0;
+
+        // transfer
+        require(token.transfer(msg.sender, amount), Errors.ERC20_ERR);
+    }
+
+    ////////////////////////
+    // Collateral Management
+    ////////////////////////
+
+    // Stakes the user for the provided batch index
+    function stake(address who, uint256 batchIndex) public returns (bool) {
+        // only callable by the state commitment chain in the `appendBatch` call
+        require(msg.sender == ovmCanonicalStateCommitmentChain);
+
+        // `batchIndex` MUST BE a strictly increasing number in the state commitment chain
+        // this check can be removed if that invariant is guaranteed
+        require(sequencers[batchIndex] == address(0), "already staked for this batch");
+
+        // lock up the collateral
+        require(bonds[who] >= requiredCollateral, Errors.NOT_ENOUGH_COLLATERAL);
+        bonds[who] -= requiredCollateral;
+
+        // store the sequencer's address as the proposer of the provided batch idx
+        sequencers[batchIndex] = who;
+
+        return true;
+    }
+
+    /// Sequencers call this function to post collateral which will be used for
+    /// the `appendBatch` call
+    function deposit(uint256 amount) public {
+        require(
+            token.transferFrom(msg.sender, address(this), amount),
+            Errors.ERC20_ERR
+        );
+
+        // This cannot overflow
+        bonds[msg.sender] += amount;
+    }
+
+    /// Sequencers call this function to withdraw collateral that they were able
+    /// to reclaim as a result of no fraud proof being initialized for their batch
+    function withdraw(uint256 amount) public {
+        require(bonds[msg.sender] >= amount);
+        bonds[msg.sender] -= amount;
+        require(
+            token.transfer(msg.sender, amount),
+            Errors.ERC20_ERR
+        );
+    }
+
+    /// Sets the required collateral for posting a state root
+    /// Callable only by the contract's deployer.
+    function setRequiredCollatreal(uint256 newValue) public {
+        require(newValue > requiredCollateral, Errors.LOW_VALUE);
```

Comment:
Should we limit this by some % increase per update?

Reply:
Yeah sure, whatever you think is an appropriate % value. @ben-chain?